lib/daru/vector.rb

```diff
@@ -1481,13 +1481,16 @@ def prepare_bootstrap(estimators)
 
     # Note: To maintain sanity, this _MUST_ be the _ONLY_ place in daru where the
     # @dtype variable is set and the underlying data type of vector changed.
-    def cast_vector_to dtype, source=nil, nm_dtype=nil
+    def cast_vector_to dtype, source=nil, nm_dtype=nil # rubocop:disable Style/CyclomaticComplexity
       source = @data.to_a if source.nil?
 
       new_vector =
         case dtype
         when :array   then Daru::Accessors::ArrayWrapper.new(source, self)
-        when :nmatrix then Daru::Accessors::NMatrixWrapper.new(source, self, nm_dtype)
+        when :nmatrix then
+          # To avoid arrays with nils throwing TypeError for :int32 default nm_dtype
+          nm_dtype = :object if nm_dtype.nil? && source.any?(&:nil?)
```

Comment:
Sorry I'm probably not explaining it well.

I'm thinking the initialiser would have to look something like this:

```
def initialize vector, context, nm_dtype=:int32
          # To avoid arrays with nils throwing TypeError for :int32 default nm_dtype
          if nm_dtype == :int32 && vector.any?(&:nil?)
            @nm_dtype = :object
          else
            @nm_dtype = @data.dtype
          end
          @size = vector.size
          @data = NMatrix.new [@size*2], vector.to_a, dtype: nm_dtype
          @context = context
          # init with twice the storage for reducing the need to resize
        end
```

Which means you're doing the `vector.any?(&:nil?)` part even for the default case where no `nm_dtype` is passed, which seems wrong to me.

What do you think?